proposals/NNNN-hlsl-matrix-type.md

```diff
@@ -0,0 +1,136 @@
+* Status: **Design In Progress**
+
+* Issues:
+  [#88060](https://github.com/llvm/llvm-project/issues/88060)
+
+## Introduction
+
+LLVM matrix support differs in some key ways from the features and limiations
+of HLSL matrices. This proposes inheriting from the existing matrix type that
+will be enabled in HLSL by default and will impose these limits and allow the
+additional capabilities.
+
+## Motivation
+
+Whether transforming geometric coordinates or adapting weights for machine
+learning, matrix support is crucial to the existing and future utility of HLSL.
+Fortunately, Clang and LLVM support an extension that provides
+the fundamentals needed for HLSL matrix support.
+However, it is too permissive in the allowable dimensions
+and lacking in various other ways that the matrices can be used.
+
+HLSL matrices in DXC have the following features that Clang matrices lack:
+
+* Matrix sizes are limited to dimensions of 1-4 inclusive
+* Initialization using vectors, scalars or a single splatted scalar.
+* (row|column)_major qualifiers determine orientation
+* `#pragma pack_matrix([row|col}_major)` changes the default orientation for
+   matrices declared below that line.
+* Orientation-appropriate column or row access using a single subscript operator
+* Zero-based ( `._m<row><col>`) and one-based (`._<row><col>`.) point accessors
+* Support for all piecewise comparison operators
+* Support for all piecewise bitwise operators for relevant types
+* Piecewise division by a scalar.
+* Boolean matrices with full operator support
+* Implicit element type conversions allowing operations on matrices of different
+  element types.
+* When both destination dimensions are less than or equal to the source, 
+  matrices can be implicitly truncated to smaller sizes with a warning.
+* Usable as resource types
+* Usable as entry point semantics
+* Matrix multiplication with compatible vectors
+* Resolve overloaded functions taking matrix parameters.
+
+Since matrices are represented as vectors,
+the following limitations of Clang vectors may also be relevant
+
+* Operators don't accept vectors with different element types
+* Boolean vectors are a different type incompatible with other vector types
+* Boolean vectors can only be used as operands to logical operators
+* Matrices and vectors have evaluation kind TEK_Scalar
+
+## Proposed solution
+
+Leverage Clang matrix support by creating HLSL-specific
+matrix types that inherit from the existing matrix extension types.
+These types will have methods enabling the additional features and
+supporting code will get variants to support the additional capability.
+
+### Leveraging Existing Support
+
+The Clang matrix extension provides us with a strong foundation on which to
+build HLSL matrices.
+Existing support gives us the ability to:
+
+* Declare matrices in different scopes (local, global, param, templates, etc)
+* Pass matrices as parameters
+* Return from functions
+* Lower to vector representations
```

Comment:
We might want to keep matrix information for SPIRV which has built-in matrix type.